v2/sourcedb-to-spanner/src/main/java/com/google/cloud/teleport/v2/source/reader/io/jdbc/dialectadapter/dialectadapter/dialectadapter/mysql/MysqlDialectAdapter.java

```diff
@@ -0,0 +1,208 @@
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.google.cloud.teleport.v2.source.reader.io.jdbc.dialectadapter.dialectadapter.dialectadapter.mysql;
+
+import com.google.cloud.teleport.v2.source.reader.io.exception.exception.RetriableSchemaDiscoveryException;
+import com.google.cloud.teleport.v2.source.reader.io.exception.exception.SchemaDiscoveryException;
+import com.google.cloud.teleport.v2.source.reader.io.jdbc.dialectadapter.dialectadapter.dialectadapter.DialectAdapter;
+import com.google.cloud.teleport.v2.source.reader.io.schema.SourceSchemaReference;
+import com.google.cloud.teleport.v2.spanner.migrations.schema.SourceColumnType;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.re2j.Pattern;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.SQLNonTransientConnectionException;
+import java.sql.SQLTransientConnectionException;
+import javax.sql.DataSource;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Adapter for MySql dialect of JDBC databases. */
+public final class MysqlDialectAdapter implements DialectAdapter {
+  private final MySqlVersion mySqlVersion;
+
+  private static final Logger logger = LoggerFactory.getLogger(MysqlDialectAdapter.class);
+
+  public MysqlDialectAdapter(MySqlVersion mySqlVersion) {
+    this.mySqlVersion = mySqlVersion;
+  }
+
+  /**
+   * @param dataSource Provider for JDBC connection.
+   * @param sourceSchemaReference Source database name and (optionally namespace)
+   * @param tables Tables to migrate.
+   * @return source table schema.
+   * @throws SchemaDiscoveryException
+   * @throws RetriableSchemaDiscoveryException
+   *     <p><b>Note:</b>
+   *     <p><b>Implementation Detail:</b>There is a choice to read from information schema table via
+   *     a prepared statement, or read the same information through jdbc metaadata calls. The major
+   *     part of parasing and mapping complexity is same in either route. We choose the route of
+   *     reading information schema table directly, since:
+   *     <ol>
+   *       <li>With all other things equal, using a prepared statement to read information schema
+   *           would be faster than making individual calls.
+   *       <li>Various drivers have different properties that can affect the search scope of the
+   *           calls.
+   *     </ol>
+   */
+  @Override
+  public ImmutableMap<String, ImmutableMap<String, SourceColumnType>> discoverTableSchema(
```

Comment:
Is it in plans to also expose additional schematic information - specifically foreign key constraints?

Reply:
In future yes. The dialect adapter can expose this information. Also for string splitting, we will need to provide range queries to JDBCIO (by default it always assumes an increasing order of the index), these range query prepare statements can be created by the dialect adapter.